### PR TITLE
Muelu: lumped diag device

### PR DESCRIPTION
**Pull request by lucbv**
@trilinos/muelu 

## Motivation
The current implementation of the lumped diagonal algorithm creates a host copy of the matrix which is very expensive compared to the computational cost of the algorithm.
To fix this performance issue a new code path that performs the lumping operation on device has been implemented.

## Stakeholder Feedback
This has resolved one of the performance bottlenecks

## Testing
This was implemented and tested on Weaver, proper testing has been included in this PR.

### CODE REVIEW

**Review by jhux2 — APPROVED** [review]
Lgtm, thanks @lucbv!

File: packages/muelu/src/Utils/MueLu_UtilitiesBase_decl.hpp [hunk 1]
```diff
@@ -300,52 +300,136 @@ namespace MueLu {
...
-            else {
-              if(TST::magnitude(diagVals[i]) > tol)
-                diagVals[i] = one / diagVals[i];
-              else {
-                diagVals[i] = valReplacement;
-              }
-            }
-          }
-        }
+	if(rowMap->lib() == Xpetra::UnderlyingLib::UseTpetra) {
```
Comment (cgcgcg) [inline packages/muelu/src/Utils/MueLu_UtilitiesBase_decl.hpp:304]: Why is the distinction between Tpetra and Epetra needed? Epetra does support the local matrix stuff, no?

Comment (jhux2) [reply packages/muelu/src/Utils/MueLu_UtilitiesBase_decl.hpp:304]: One can get a localDeviceView?

Comment (cgcgcg) [reply packages/muelu/src/Utils/MueLu_UtilitiesBase_decl.hpp:304]: https://github.com/trilinos/Trilinos/blob/78cf276df2e4a91296100d0ee25c31d93d18e976/packages/xpetra/src/CrsMatrix/Xpetra_EpetraCrsMatrix.hpp#L1230-L1237

Comment (lucbv) [reply packages/muelu/src/Utils/MueLu_UtilitiesBase_decl.hpp:304]: I honestly did not want to have to check and do no think that there is a real benefit to not run the old code when Epetra is being used...

**Comment by cgcgcg** [comment]
Different question: where is the input argument `tol` of the method used?

**Comment by jhux2** [comment]
@lucbv Compile errors fixed, now just one test failure:

<details>

```
929. Utilities_std_complex0double0_int_longlong_Kokkos_Compat_KokkosCudaWrapperNode_GetLumpedDiagonal_UnitTest ... 
 
 Check: rel_err(diag[idx], diag_ref[idx])
        = rel_err((6,0), (6,0)) = 0
          <= tol = 2.22045e-14 : passed
 
 Check: rel_err(diag[idx], diag_ref[idx])
        = rel_err((7,0), (7,0)) = 0
          <= tol = 2.22045e-14 : passed
 
 Check: rel_err(diag[idx], diag_ref[idx])
        = rel_err((7,0), (7,0)) = 0
          <= tol = 2.22045e-14 : passed
 
 Check: rel_err(diag[idx], diag_ref[idx])
        = rel_err((7,0), (7,0)) = 0
          <= tol = 2.22045e-14 : passed
 
 Check: rel_err(diag[idx], diag_ref[idx])
        = rel_err((6.5,0), (6.5,0)) = 0
          <= tol = 2.22045e-14 : passed
 
 Check: rel_err(diag[idx], diag_ref[idx])
        = rel_err((2,0), (2,0)) = 0
          <= tol = 2.22045e-14 : passed
 
 Check: rel_err(diag[idx], diag_ref[idx])
        = rel_err((0.7,0), (0.7,0)) = 0
          <= tol = 2.22045e-14 : passed
 
 Check: rel_err(diag[idx], diag_ref[idx])
        = rel_err((8,0), (8,0)) = 0
          <= tol = 2.22045e-14 : passed
 
 Check: rel_err(diag[idx], diag_ref[idx])
        = rel_err((8,0), (8,0)) = 0
          <= tol = 2.22045e-14 : passed
 
 Check: rel_err(diag[idx], diag_ref[idx])
        = rel_err((0.166667,0), (0.166667,0)) = 0
          <= tol = 2.22045e-14 : passed
 
 Check: rel_err(diag[idx], diag_ref[idx])
        = rel_err((0.142857,0), (0.142857,0)) = 0
          <= tol = 2.22045e-14 : passed
 
 Check: rel_err(diag[idx], diag_ref[idx])
        = rel_err((0.142857,0), (0.142857,0)) = 0
          <= tol = 2.22045e-14 : passed
 
 Check: rel_err(diag[idx], diag_ref[idx])
        = rel_err((0.142857,0), (0.142857,0)) = 0
          <= tol = 2.22045e-14 : passed
 
 Check: rel_err(diag[idx], diag_ref[idx])
        = rel_err((0.153846,0), (0.153846,0)) = 0
          <= tol = 2.22045e-14 : passed
 
 Check: rel_err(diag[idx], diag_ref[idx])
        = rel_err((0.25,0), (0.25,0)) = 0
          <= tol = 2.22045e-14 : passed
 
 Check: rel_err(diag[idx], diag_ref[idx])
        = rel_err((1.42857,0), (1.42857,0)) = 0
          <= tol = 2.22045e-14 : passed
 
 Check: rel_err(diag[idx], diag_ref[idx])
        = rel_err((0.125,0), (0.125,0)) = 0
          <= tol = 2.22045e-14 : passed
 
 Check: rel_err(diag[idx], diag_ref[idx])
        = rel_err((0.125,0), (0.125,0)) = 0
          <= tol = 2.22045e-14 : passed
 
 Check: rel_err(diag[idx], diag_ref[idx])
        = rel_err((0.166667,0), (0.166667,0)) = 0
          <= tol = 2.22045e-14 : passed
 
 Check: rel_err(diag[idx], diag_ref[idx])
        = rel_err((0.142857,0), (0.142857,0)) = 0
          <= tol = 2.22045e-14 : passed
 
 Check: rel_err(diag[idx], diag_ref[idx])
        = rel_err((0.142857,0), (0.142857,0)) = 0
          <= tol = 2.22045e-14 : passed
 
 Check: rel_err(diag[idx], diag_ref[idx])
        = rel_err((0.142857,0), (0.142857,0)) = 0
          <= tol = 2.22045e-14 : passed
 
 Check: rel_err(diag[idx], diag_ref[idx])
        = rel_err((0.153846,0), (0.153846,0)) = 0
          <= tol = 2.22045e-14 : passed
 
 Check: rel_err(diag[idx], diag_ref[idx])
        = rel_err((0.25,0), (0.25,0)) = 0
          <= tol = 2.22045e-14 : passed
 
 Check: rel_err(diag[idx], diag_ref[idx])
        = rel_err((1.42857,0), (0,0)) = 1
          <= tol = 2.22045e-14 : FAILED
 
 Check: rel_err(diag[idx], diag_ref[idx])
        = rel_err((0.125,0), (0.125,0)) = 0
          <= tol = 2.22045e-14 : passed
 
 Check: rel_err(diag[idx], diag_ref[idx])
        = rel_err((0.125,0), (0.125,0)) = 0
          <= tol = 2.22045e-14 : passed
 
 Check: rel_err(diag[idx], diag_ref[idx])
        = rel_err((0.166667,0), (0.166667,0)) = 0
          <= tol = 2.22045e-14 : passed
 
 Check: rel_err(diag[idx], diag_ref[idx])
        = rel_err((0.142857,0), (0.142857,0)) = 0
          <= tol = 2.22045e-14 : passed
 
 Check: rel_err(diag[idx], diag_ref[idx])
        = rel_err((0.142857,0), (0.142857,0)) = 0
          <= tol = 2.22045e-14 : passed
 
 Check: rel_err(diag[idx], diag_ref[idx])
        = rel_err((0.142857,0), (0.142857,0)) = 0
          <= tol = 2.22045e-14 : passed
 
 Check: rel_err(diag[idx], diag_ref[idx])
        = rel_err((0.153846,0), (0.153846,0)) = 0
          <= tol = 2.22045e-14 : passed
 
 Check: rel_err(diag[idx], diag_ref[idx])
        = rel_err((0,0), (0,0)) = 0
          <= tol = 2.22045e-14 : passed
 
 Check: rel_err(diag[idx], diag_ref[idx])
        = rel_err((1.42857,0), (1.42857,0)) = 0
          <= tol = 2.22045e-14 : passed
 
 Check: rel_err(diag[idx], diag_ref[idx])
        = rel_err((0.125,0), (0.125,0)) = 0
          <= tol = 2.22045e-14 : passed
 
 Check: rel_err(diag[idx], diag_ref[idx])
        = rel_err((0.125,0), (0.125,0)) = 0
          <= tol = 2.22045e-14 : passed
 
 Check: rel_err(diag[idx], diag_ref[idx])
        = rel_err((0.166667,0), (0.166667,0)) = 0
          <= tol = 2.22045e-14 : passed
 
 Check: rel_err(diag[idx], diag_ref[idx])
        = rel_err((0.142857,0), (0.142857,0)) = 0
          <= tol = 2.22045e-14 : passed
 
 Check: rel_err(diag[idx], diag_ref[idx])
        = rel_err((0.142857,0), (0.142857,0)) = 0
          <= tol = 2.22045e-14 : passed
 
 Check: rel_err(diag[idx], diag_ref[idx])
        = rel_err((0.142857,0), (0.142857,0)) = 0
          <= tol = 2.22045e-14 : passed
 
 Check: rel_err(diag[idx], diag_ref[idx])
        = rel_err((0.153846,0), (0.153846,0)) = 0
          <= tol = 2.22045e-14 : passed
 
 Check: rel_err(diag[idx], diag_ref[idx])
        = rel_err((0,0), (0,0)) = 0
          <= tol = 2.22045e-14 : passed
 
 Check: rel_err(diag[idx], diag_ref[idx])
        = rel_err((1.42857,0), (42,0)) = 0.965986
          <= tol = 2.22045e-14 : FAILED
 
 Check: rel_err(diag[idx], diag_ref[idx])
        = rel_err((0.125,0), (0.125,0)) = 0
          <= tol = 2.22045e-14 : passed
 
 Check: rel_err(diag[idx], diag_ref[idx])
        = rel_err((0.125,0), (0.125,0)) = 0
          <= tol = 2.22045e-14 : passed
 [FAILED]  (0.00166 sec) Utilities_std_complex0double0_int_longlong_Kokkos_Compat_KokkosCudaWrapperNode_GetLumpedDiagonal_UnitTest
 Location: ../Trilinos/packages/muelu/test/unit_tests/Utilities.cpp:282

```

</details>

**Comment by jhux2** [comment]
> Different question: where is the input argument `tol` of the method used?

Here, for instance:

https://github.com/trilinos/Trilinos/blob/77b50ad9f4ae5a88aa60fa8855c59b7ec39e7b6b/packages/muelu/src/Utils/MueLu_UtilitiesBase_decl.hpp#L365

**Comment by cgcgcg** [comment]
Oh. Yeah. That value only gets overwritten when `useAvgAbsDiagVal == true`. Gotcha.

**Comment by lucbv** [comment]
@jhux2 I rebuilt three times on weaver and was not able to reproduce that numerical error

**Comment by jhux2** [comment]
> I rebuilt three times on weaver and was not able to reproduce that numerical error

@lucbv I'm have a build going to try to reproduce on one of the ascicgpu machines.  I'm using these instructions: https://github.com/trilinos/Trilinos/wiki/Reproducing-PR-Testing-Errors.

**Comment by jhux2** [comment]
> I'm have a build going to try to reproduce on one of the ascicgpu machines.

Ok, I am able to reproduce locally.

**Review by jhux2 — APPROVED** [review]
best. commits. ever.

**Pull request merged**